SDR-Python bindings - WIP

ctrl-z-9000-times opened the pull request:
See issue #211 

Progress so far:
- Implementation finished, pybind11 is really easy to use :)
- Python can read all data with NO copy.
- Python can write dense data with NO copy, by modifying inplace and using method `sdr.setDenseInplace()`
- Write flatSparse & sparse data from Python incurs a copy.

TODO:
- [x] Unit tests
- [x] Documentation
- [x] SDR_Proxy bindings
- [x] SDR_Proxy docs
- [x] SDR_Proxy unit tests
- [x] SDR_Metrics fixups [PR #232]
- [x] SDR_Metrics bindings
- [x] SDR_Metrics docs
- [x] SDR_Metrics unit tests

dkeeney commented on bindings/py/cpp_src/CMakeLists.txt:
```diff
@@ -41,6 +41,7 @@ message(STATUS "Configuring Python interface")
 #
 set(src_py_algorithms_files
     bindings/algorithms/algorithm_module.cpp
+    bindings/algorithms/py_SDR.cpp
```
Very good 👍

breznak replied:
MSVC branch merged, this should be fixed.

breznak commented on bindings/py/tests/algorithms/SDR_test.py:
```diff
@@ -0,0 +1,242 @@
+        A = SDR((103,))
+        B = SDR((100, 100, 1))
+        assert(tuple(B.dimensions) == (100, 100, 1))
+        C = SDR((2, 4, 5, 1,1,1,1, 3))
```
this is a bit confusing, what is it, initialization by given set of active bits? Add a doc please

ctrl-z-9000-times replied:
Ok, In the next commit I will change this to:
```
        # Test crazy dimensions, also test keyword arguments.
        C = SDR(dimensions = (2, 4, 5, 1,1,1,1, 3) )
```

breznak replied:
so it's an 8-dim SDR?

breznak commented on bindings/py/tests/algorithms/SDR_test.py:
```diff
@@ -0,0 +1,242 @@
+    def testOverlap(self):
+        A = SDR((103,))
+        B = SDR((103,))
+        assert(A.overlap(B) == 0)
```
does `overlap` alter A, or B? If not, I'd prefer name `getOverlap` for both py, c++

ctrl-z-9000-times replied:
`A.overlap(B)` does not change the value of either `A` or `B` however it does require that both be non-constant.

breznak replied:
I'd say get* would be more natural in that case

breznak replied:
👍 thanks

breznak commented on src/nupic/utils/StringUtils.cpp:
```diff
@@ -394,17 +394,17 @@ bool StringUtils::toIntListNoThrow(const std::string &s, std::vector<Int> &list,
 
 //--------------------------------------------------------------------------------
-std::shared_ptr<Byte> StringUtils::toByteArray(const std::string &s,  Size bitCount) {
+std::shared_ptr<char> StringUtils::toByteArray(const std::string &s,  Size bitCount) {
```
not in this PR, but could we change these to `string` return types? Even if that breaks compat, I think it would be simpler code. @dkeeney?

ctrl-z-9000-times replied:
Ok, I just now reverted this change.  The reason I replaced Byte with char was to make pybind11 happy with the type conversions, which I have since fixed in a different way.

ctrl-z-9000-times commented:
I don't know why pickle seg-faults, I can't reproduce it on my machine.

dkeeney commented:
Yes, that has been on my list for a while.  I don't know where to start either.
My suspect is that from the C++ side we start a pickle of some .py class and somewhere it calls back into the C++ where we don't pick it up and give it a serialization of the C++ part. (.py --> C++) interface.
```
  C++ --> .py  --> C++
```

breznak commented on bindings/py/cpp_src/bindings/algorithms/py_SDR.cpp:
```diff
@@ -255,11 +255,30 @@ special, it is replaced with the system time.  The default seed is 0.)",
 
-        py::class_<SDR_Proxy, SDR> py_Proxy(m, "SDR_Proxy");
+        py::class_<SDR_Proxy, SDR> py_Proxy(m, "SDR_Proxy",
+R"(SDR_Proxy presents a view onto an SDR.
```
what is this R" supposted to be? some formating?

ctrl-z-9000-times replied:
It's the start of a multi-line string.
* Starts with `R"(`
* Ends with `)"`  Yes, the parenthesis go inside of the quotes
* Anything inside, arbitrary length, no need to backslash escape line endings.

I think they added this is C++11? Its great, right?

ctrl-z-9000-times commented:
Previously I had said:

>    I plan on added a convenience to the SDR class:
`SDR( UInt dimensions ) : SDR( {dimensions} ) {};`

This special case would allow users to make a 1-D SDR without making a vector of dimensions.  For python I *will* add this feature.

However in C++ this causes the compile error:
```
error: braces around scalar initializer [-Werror,-Wbraced-scalar-init]
  SDR foobar( {1234} );
```
So it appears that we can't have it both ways, we need to pick between:
`SDR( 1234 )` or `SDR( {1234} )` and only allow one way to do this.  For C++ I'm going to only have the vector initializer since the purpose of this convenience is to relax the rules about when to use {}'s, not add more rules about {}'s.

dkeeney commented:
> So it appears that we can't have it both ways

No problem.  I think I do that in only one place anyway.

breznak commented on bindings/py/tests/SDR_Metrics_test.py:
```diff
@@ -0,0 +1,123 @@
+
+class SdrMetricsTest(unittest.TestCase):
+    def testSparsityExample(self):
+        A = SDR( dimensions = 1000 )
```
nit, but imho you might consider also removing the SDR(int) binding. Keep it only vector/list, I was still wondering here if it's a 1000 dimensional SDR, or 1D with width 1000. SDR(dimensions = [1000]) is imho much clearer.

breznak commented on bindings/py/tests/SDR_test.py:
```diff
@@ -0,0 +1,318 @@
+
+    def testLostSDR(self):
+        # You need to keep a reference to the SDR, since SDR class does not use smart pointers.
+        B = SDR_Proxy(SDR((1000,)))
```
not for this PR, but this is tricky behavior. But probably having SDR_PRoxy manage collection of pointers would be as nasty.

ctrl-z-9000-times replied:
The purpose of this unit-test is the ensure that if/when the parent SDR is lost, the proxy doesn't crash & die.  The Proxy is notified when the parent is lost, and the proxy invalidates itself so this will never segfault, just throws an exception.

What's more worrying is:
```
python3
>>> A = SDR([1000])
>>> memory = A.dense
>>> del A
>>> memory -> SDR is gone, dense data is freed, but numpy array survives
```

Or more simply `SDR([1000]).dense` will allow access to memory that has been freed.

breznak commented on src/nupic/ntypes/SdrMetrics.hpp:
```diff
@@ -127,6 +123,10 @@ class _SDR_MetricsHelper {
         callback( data, 1.0f / std::min( period_, (UInt) ++samples_ ));
     }
+
+    ~_SDR_MetricsHelper() {
```
nit, please use This_ format rather. And If you are at it: `SDRMetricsHelper_`

ctrl-z-9000-times replied:
I just changed it to `SDR_MetricsHelper_`.

breznak commented on src/nupic/types/Types.hpp:
```diff
@@ -44,7 +44,11 @@ namespace nupic {
 /**
  * Represents a 8-bit byte.
  */
-typedef int8_t Byte;
```
what is the advantage to original `char` here? What is this Byte used to? (uint_8_least/fast_t ...?)

ctrl-z-9000-times replied:
Pybind11 automatically converts based on data type.  `char` array converts to/from unicode strings, `int8` converts to/from integers.

ctrl-z-9000-times commented:
Windows doesn't define the same C++ types, at least not in the same way as on unix.  This will take some time to sort out.

breznak commented:
> Windows doesn't define the same C++ types

I think the int8 are not mandated by the standard

dkeeney commented:
Some older windows libraries defined Byte as a primitive type (unsigned char) used as a number.
But nupic::Byte is a signed char.   I would rather see it changed to unsigned int8 so it is obvious that it is not a character but I expect it would break a lot of compiles at this point.

I do have MSVC getting confused when it sees the typedef for nupic:Byte so I have to restart my IDE and it seems to work again.

ctrl-z-9000-times commented:
I fixed the issue with Windows & Data types.

The solution was to change the pybind code from `vector<Byte>` to `numpy_array<Byte>`.  This is a great solution because the SDR also gives a numpy array to python, so if you do:
`sdr.dense = sdr.dense`
Then the SDR can detect that you've given it the same data buffer which it gave you.  Then it can assign its data inplace without a copy operation.  Use this one liner to update an SDR's data inplace instead of calling the method `sdr.setDenseInplace()` which I removed.

ctrl-z-9000-times commented on README.md:
````diff
@@ -82,12 +82,16 @@ Fork or download the HTM-Community Nupic.cpp repository from https://github.com/
 ```
     cd to-repository-root
-    python setup.py install --user --prefix=
+    python setup.py install --user --force --prefix=
````
IIRC: --user and --prefix are mutually exclusive.  --user installs to a hidden folder in your home directory.  --prefix takes a CLI argument which is where to install to.

dkeeney replied:
Ah, then maybe we can safely drop the --prefix= clause.

ctrl-z-9000-times replied:
I agree.  `--user` should cover 99% of the use cases.  --prefix is for people with linux & a custom file organization.  Furthermore the `--prefix` comes from the python distutils package which is well documented: `setup.py --help` shows this option.

breznak replied:
the `--prefix=` had some meaning even with --user, it was a hack. Imho for SWIG. Ok to let go.

ctrl-z-9000-times commented on src/nupic/ntypes/Sdr.hpp:
```diff
@@ -255,14 +255,23 @@ class SparseDistributedRepresentation : public Serializable
     SparseDistributedRepresentation( const vector<UInt> dimensions )
         { initialize( dimensions ); }
 
+    /**
```
`SDR( 10 )` instead of `SDR( { 10 } )`
This only works for Python.  In C++ this convenience caused too many compiler warnings.

dkeeney replied:
No problem, I can create the vector.

ctrl-z-9000-times commented on bindings/py/cpp_src/bindings/algorithms/py_SDR.cpp:
```diff
@@ -0,0 +1,319 @@
+    A.sparse =  ([1, 1, 2], [0, 1, 2])
+    B.sparse -> ([2, 2, 5], [0, 1, 0])
+
+SDR_Proxy does not support pickle)");
```
My comment is wrong, SDR_Proxy does support saving pickles, it just doesn't support loading pickles.  SDR_Proxies are supposed to be read only views on the parent SDR, and loading would change the parent so it is not allowed.  Instead the users must load the parent SDR.

breznak commented on README.md:
````diff
@@ -82,12 +82,16 @@ Fork or download the HTM-Community Nupic.cpp repository from https://github.com/
 ```
     cd to-repository-root
-    python setup.py install --user --prefix=
+    python setup.py install --user --force
````
do we eat our dog food? is this used in CI?

ctrl-z-9000-times replied:
Travis ci does not use this. I dont know about yhe other CIs.  I personally use these flags.
